spark/src/test/scala/org/apache/spark/sql/delta/IdentityColumnSuite.scala

```diff
@@ -124,6 +130,251 @@ trait IdentityColumnSuiteBase extends IdentityColumnTestUtils {
+            IdentityColumnSpec(
+              GeneratedByDefault,
+              startsWith = Some(1),
+              incrementBy = Some(1),
+              colName = "id1"
+            ),
+            IdentityColumnSpec(
+              GeneratedAlways,
+              startsWith = Some(1),
+              incrementBy = Some(1),
+              colName = "id2"
+            ),
+            IdentityColumnSpec(
+              GeneratedAlways,
+              startsWith = Some(1),
+              incrementBy = Some(1),
+              colName = "id3"
+            ),
+            TestColumnSpec(colName = "value", dataType = IntegerType)
+          )
+        )
+      }.filter { e =>
+        e.tags.get("opType").exists(_ == IdentityColumn.opTypeDefinition)
+      }
+      assert(eventsDefinition.size == 1)
+      assert(JsonUtils.fromJson[Map[String, String]](eventsDefinition.head.blob)
+        .get("numIdentityColumns").exists(_ == "3"))
+
+      val eventsWrite = Log4jUsageLogger.track {
+        sql(s"INSERT INTO $tblName (id1, value) VALUES (1, 10), (2, 20)")
+      }.filter { e =>
+        e.tags.get("opType").exists(_ == IdentityColumn.opTypeWrite)
+      }
+      assert(eventsWrite.size == 1)
+      val data = JsonUtils.fromJson[Map[String, String]](eventsWrite.head.blob)
+      assert(data.get("numInsertedRows").exists(_ == "2"))
+      assert(data.get("generatedIdentityColumnNames").exists(_ == "id2,id3"))
+      assert(data.get("generatedIdentityColumnCount").exists(_ == "2"))
+      assert(data.get("explicitIdentityColumnNames").exists(_ == "id1"))
+      assert(data.get("explicitIdentityColumnCount").exists(_ == "1"))
+    }
+  }
+
+  test("reading table should not see identity column properties") {
+    def verifyNoIdentityColumn(id: Int, f: () => Dataset[_]): Unit = {
+      assert(!ColumnWithDefaultExprUtils.hasIdentityColumn(f().schema), s"test $id failed")
+    }
+
+    withTable(tblName) {
+      createTable(
+        tblName,
+        Seq(
+          IdentityColumnSpec(GeneratedByDefault),
+          TestColumnSpec(colName = "part", dataType = LongType),
+          TestColumnSpec(colName = "value", dataType = StringType)
+        ),
+        partitionedBy = Seq("part")
+      )
+
+      sql(
+        s"""
+           |INSERT INTO $tblName (part, value) VALUES
+           |  (1, "one"),
+           |  (2, "two"),
+           |  (3, "three")
+           |""".stripMargin)
+      val path = DeltaLog.forTable(spark, TableIdentifier(tblName)).dataPath.toString
+
+      val commands: Map[Int, () => Dataset[_]] = Map(
+        1 -> (() => spark.table(tblName)),
+        2 -> (() => sql(s"SELECT * FROM $tblName")),
+        3 -> (() => sql(s"SELECT * FROM delta.`$path`")),
+        4 -> (() => spark.read.format("delta").load(path)),
+        5 -> (() => spark.read.format("delta").table(tblName)),
+        6 -> (() => spark.readStream.format("delta").load(path)),
+        7 -> (() => spark.readStream.format("delta").table(tblName))
+      )
```

Comment:
We either don't need the indices here or we just make it the parameter (Int) => Dataset[]. They're just used for "test number". Let's simplify this.